odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,556 @@
+
+    """Abstract operator for point-wise tensor field manipulations.
+
+    A point-wise operator acts on a space of vector or tensor fields,
+    i.e. a power space ``X^d`` of a discretized function space ``X``.
+    Its range is the power space ``X^k`` with a possibly different
+    number ``k`` of components.
+
+    For example, if ``X`` is a `DiscreteLp` space, then
+    ``ProductSpace(X, d)`` is a valid domain for any positive integer
+    ``d``.
+
+    Currently, only vector fields, i.e. one-dimensional products of
+    ``X``, are supported.
+
+    See also
+    --------
+    ProductSpace
+    """
+
+    def __init__(self, domain, range, linear=False):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        domain, range : {`ProductSpace`, `Discretization`}
+            Spaces of vector fields between which the operator maps.
+            They have to be either power spaces of the same base space
+            ``X`` or the base space itself (only one of them).
+        linear : `bool`, optional
+            If `True`, assume that the operator is linear.
+        """
+        if isinstance(domain, ProductSpace):
+            if not domain.is_power_space:
+                raise TypeError('domain {!r} is not a power space.'
+                                ''.format(domain))
+
+            if not isinstance(domain[0].uspace, FunctionSpace):
+                raise TypeError('domain base space {!r} is not a function '
+                                'space discretization.'.format(domain[0]))
+
+            dom_base = domain[0]
+        else:
+            dom_base = domain
+
+        if isinstance(range, ProductSpace):
+            if not range.is_power_space:
+                raise TypeError('range {!r} is not a power space.'
+                                ''.format(range))
+
+            if not isinstance(range[0].uspace, FunctionSpace):
+                raise TypeError('range base space {!r} is not a function '
+                                'space discretization.'.format(range[0]))
+
+            ran_base = range[0]
+        else:
+            ran_base = range
+
+        if dom_base != ran_base:
+            raise ValueError('domain and range have different base spaces '
+                             '({!r} != {!r}).'
+                             ''.format(domain[0], range[0]))
+
+        super().__init__(domain=domain, range=range, linear=linear)
+        self._base_space = dom_base
+
+    @property
+    def base_space(self):
+        """The base space ``X`` of this operator's domain and range."""
+        return self._base_space
+
+
+class PointwiseNorm(PointwiseOperator):
+
+    """Take the point-wise norm of a tensor field.
+
+    This operator takes the (weighted) ``p``-norm
+
+        ``||F(x)|| = [ sum_j( w_j * |F_j(x)|^p ) ]^(1/p)``
+
+    for ``p`` finite and
+
+        ``||F(x)|| = max_j( w_j * |F_j(x)| )``
+
+    for ``p = inf``, where ``F`` is a tensor field. This implies that
```

Comment:
Definitions will change for higher order tensor fields, so probably sticking with vector fields it should be for now.